packages/Search/src/DTK_LinearBVH_fwd.hpp

```diff
@@ -0,0 +1,139 @@
+
+    KOKKOS_INLINE_FUNCTION
+    size_type size() const { return _leaf_nodes.extent( 0 ); }
+
+    KOKKOS_INLINE_FUNCTION
+    bool empty() const { return size() == 0; }
+
+    KOKKOS_INLINE_FUNCTION
+    bounding_volume_type bounds() const
+    {
+        // NOTE should default constructor initialize to an invalid geometry?
+        if ( empty() )
+            return bounding_volume_type();
+        return getBoundingVolume( this->getRoot() );
+    }
+
+    template <typename Predicates, typename... Args>
+    inline void query( Predicates const &predicates, Args &&... args ) const
+    {
+        // FIXME lame placeholder for concept check
+        static_assert( Kokkos::is_view<Predicates>::value, "must pass a view" );
+        using Tag = typename Predicates::value_type::Tag;
+        Details::BoundingVolumeHierarchyImpl<DeviceType>::queryDispatch(
+            Tag{}, *this, predicates, std::forward<Args>( args )... );
+    }
+
+    // private:
+  public:
+    Kokkos::View<Node *, DeviceType, Kokkos::MemoryUnmanaged> getLeafNodes()
+    {
+        return _leaf_nodes;
+    }
+
+    /**
+     * Return the root node of the BVH.
+     */
+    KOKKOS_INLINE_FUNCTION Node *getRoot()
+    {
+        if ( empty() )
+            return nullptr;
+        return ( size() > 1 ? _internal_nodes : _leaf_nodes ).data();
+    }
+    KOKKOS_INLINE_FUNCTION Node const *getRoot() const
+    {
+        if ( empty() )
+            return nullptr;
+        return ( size() > 1 ? _internal_nodes : _leaf_nodes ).data();
+    }
+    /**
+     * Return the bounding volume of a node of the BVH.
+     */
+    KOKKOS_INLINE_FUNCTION Box &getBoundingVolume( Node *node )
+    {
+        return node->bounding_box;
+    }
+    KOKKOS_INLINE_FUNCTION Box const &
+    getBoundingVolume( Node const *node ) const
+    {
+        return node->bounding_box;
+    }
+    /**
+     * Return true if the node is a leaf.
+     */
+    KOKKOS_INLINE_FUNCTION bool isLeaf( Node const *node ) const
+    {
+        return ( node->children.first == nullptr );
+    }
+    /**
+     * Return the index of a leaf node.
+     */
+    // FIXME: debate on whether we only need getPermutationIndex for leaf
+    // nodes, and what to do (if we need to) for internal
+    KOKKOS_INLINE_FUNCTION int getPermutationIndex( Node const *leaf ) const
+    {
+        static_assert( sizeof( size_t ) == sizeof( Node * ),
+                       "Conversion is a bad idea if these sizes do not match" );
+        return reinterpret_cast<size_t>( leaf->children.second );
+    }
+    /**
+     * Return the offset index of the a node.
+     */
+    KOKKOS_INLINE_FUNCTION int getOffsetIndex( Node const *node ) const
+    {
+        return static_cast<int>( ( isLeaf( node )
+                                       ? ( node - _leaf_nodes.data() )
+                                       : ( node - getRoot() ) ) );
+    }
+    KOKKOS_INLINE_FUNCTION int getInternalOffsetIndex( Node const *node ) const
```

Comment:
Get rid of that member function.  I don't see how `internal_node - root` would be invalidated by any of the data changes we wanted to experiment.